zookeeper-server/src/test/java/org/apache/zookeeper/server/ConnectionMetricsTest.java

```diff
@@ -0,0 +1,228 @@
+        }
+
+        @Override
+        public boolean isValid() {
+            return true;
+        }
+
+        @Override
+        public void cancel() {
+        }
+
+        @Override
+        public int interestOps() {
+            return ops;
+        }
+
+        private int ops = OP_WRITE + OP_READ;
+
+        @Override
+        public SelectionKey interestOps(int ops) {
+            this.ops = ops;
+            return this;
+        }
+
+        @Override
+        public int readyOps() {
+            return ops;
+        }
+
+    }
+
+    private NIOServerCnxn createMockNIOCnxn() throws IOException {
+        InetSocketAddress socketAddr = new InetSocketAddress(80);
+        Socket socket = mock(Socket.class);
+        when(socket.getRemoteSocketAddress()).thenReturn(socketAddr);
+        SocketChannel sock = mock(SocketChannel.class);
+        when(sock.socket()).thenReturn(socket);
+        when(sock.read(any(ByteBuffer.class))).thenReturn(-1);
+
+        return new MockNIOServerCnxn(mock(ZooKeeperServer.class), sock, null, mock(NIOServerCnxnFactory.class), null);
+    }
+
+    @Test
+    public void testNIOConnectionDropCount() throws Exception {
+        ServerMetrics.resetAll();
+
+        NIOServerCnxn cnxn = createMockNIOCnxn();
+        cnxn.doIO(new FakeSK());
+
+        Map<String, Object> values = ServerMetrics.getAllValues();
+        Assert.assertEquals(1L, values.get("connection_drop_count"));
+    }
+
+    @Test
+    public void testNettyConnectionDropCount() throws Exception {
+        InetSocketAddress socketAddr = new InetSocketAddress(80);
+        Channel channel = mock(Channel.class);
+        when(channel.isOpen()).thenReturn(false);
+        when(channel.remoteAddress()).thenReturn(socketAddr);
+        EventLoop eventLoop = mock(EventLoop.class);
+        when(channel.eventLoop()).thenReturn(eventLoop);
+
+        ServerMetrics.resetAll();
+
+        NettyServerCnxnFactory factory = new NettyServerCnxnFactory();
+        NettyServerCnxn cnxn = new NettyServerCnxn(channel, mock(ZooKeeperServer.class), factory);
+
+        // pretend it's connected
+        factory.cnxns.add(cnxn);
+        cnxn.close();
+
+        Map<String, Object> values = ServerMetrics.getAllValues();
+        Assert.assertEquals(1L, values.get("connection_drop_count"));
+    }
+
+    @Test
+    public void testSessionlessConnectionsExpired() throws Exception {
+        ServerCnxnFactory factory = new NIOServerCnxnFactory();
+        factory.configure(new InetSocketAddress(PortAssignment.unique()), 1000);
+        factory.start();
+        int timeout = Integer.getInteger(
+                ZOOKEEPER_NIO_SESSIONLESS_CNXN_TIMEOUT, 10000);
+
+        ServerMetrics.resetAll();
+        // add two connections w/o touching them so they will expire
+        ((NIOServerCnxnFactory) factory).touchCnxn(createMockNIOCnxn());
+        ((NIOServerCnxnFactory) factory).touchCnxn(createMockNIOCnxn());
+
+        Thread.sleep(2*timeout);
```

Comment:
What about having a loop that checks for the expected value, with a 100 ms sleep?
 This will make the test faster and less flaky

Maybe we'd better have an utility method, but can we do it in a separate work

Reply:
Changed. But the tests don't run faster :(